core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/wdlparsing/CachingThreadPoolWDLParser.scala

```diff
@@ -0,0 +1,98 @@
+
+import java.util.concurrent.{Callable, ExecutorService, Executors, TimeUnit}
+
+import com.github.benmanes.caffeine.cache.Caffeine
+import com.typesafe.config.ConfigFactory
+import com.typesafe.scalalogging.LazyLogging
+import languages.wdl.draft2.WdlDraft2LanguageFactory.httpResolver
+import org.broadinstitute.dsde.rawls.RawlsException
+import scalacache.{Cache, Entry, get, put}
+import scalacache.caffeine.CaffeineCache
+import wdl.draft2.model.{WdlNamespaceWithWorkflow, WdlWorkflow}
+import wdl.draft2.parser.WdlParser.SyntaxError
+
+import scala.collection.JavaConverters._
+import scala.util.hashing.MurmurHash3
+import scala.util.{Failure, Try}
+
+
+object CachingThreadPoolWDLParser extends WDLParsing with LazyLogging {
+
+  // TODO: conf should be injected, not read directly. At least this is an object so it happens once.
+  private val conf = ConfigFactory.parseResources("version.conf").withFallback(ConfigFactory.load())
+  private val wdlParsingConf = conf.getConfig("wdl-parsing")
+
+  private val cacheMaxSize = wdlParsingConf.getInt("cache-max-size")
+  private val threadPoolSize = wdlParsingConf.getInt("parser-thread-pool-size")
+  private val threadPoolTimeout = wdlParsingConf.getInt("parser-thread-pool-timeout-seconds")
+
+  // set up cache for WDL parsing
+  /* from scalacache doc: "Note: If you’re using an in-memory cache (e.g. Guava or Caffeine) then it makes sense
+     to use the synchronous mode. But if you’re communicating with a cache over a network (e.g. Redis, Memcached)
+     then this mode is not recommended. If the network goes down, your app could hang forever!"
+   */
+  import scalacache.modes.sync._
+  private val underlyingCaffeineCache = Caffeine.newBuilder()
+    .maximumSize(cacheMaxSize)
+    .build[String, Entry[Try[WdlWorkflow]]]
+  implicit val customisedCaffeineCache: Cache[Try[WdlWorkflow]] = CaffeineCache(underlyingCaffeineCache)
+
+  // set up thread pool
+  val executorService: ExecutorService = Executors.newFixedThreadPool(threadPoolSize)
+
+  def parse(wdl: String): Try[WdlWorkflow] = {
+    val tick = System.currentTimeMillis()
+    val key = generateCacheKey(wdl)
+    logger.debug(s"<parseWDL-cache> looking up $key ...")
+    get(key) match {
+      case Some(parseResult) =>
+        val tock = System.currentTimeMillis() - tick
+        logger.debug(s"<parseWDL-cache> found cached result for $key in $tock ms.")
+        parseResult
+      case None =>
+        val miss = System.currentTimeMillis() - tick
+        logger.debug(s"<parseWDL-cache> encountered cache miss for $key in $miss ms.")
+        val parseResult = threadedParse(wdl)
+        val parsetime = System.currentTimeMillis() - tick
+        logger.debug(s"<parseWDL-cache> actively parsed WDL for $key in $parsetime ms.")
+        put(key)(parseResult)
+        val tock = System.currentTimeMillis() - tick
+        logger.debug(s"<parseWDL-cache> cached result $key in $tock ms.")
+        parseResult
+    }
+  }
```

Comment:
1) TODO: this should not cache potentially-transient errors. We want to retry those.

2) Instead of a bounded thread pool in `threadedParse` below, we could synchronize the `parse` method above, using `key` as the synchronization object. Thoughts?

Reply:
1. Ah agreed, I commented to that effect above.
2. 🤷‍♂️. It's either add synchronization or potentially do unnecessary work if >1 people load an uncached MC at the same time. I'd probably lean towards the latter (e.g. NOT adding synchronization, the way it is now). IMO this shouldn't be such an expensive operation that we need to synchronize.

Reply:
If it's true that

- it's pretty rare to keep refreshing on a broken WDL 
- it's relatively easy to get a transient failure resolving remote files due to a network blip
- the sttp fix should make things better in general

then... I think it's ok to only cache successes.

Reply:
if it is fast to parse (sub second) then I without sync is probably fine. But the longer it takes, the bigger the window for multiple requests to come in and get us in the situation this fix is trying to fix.

Reply:
now with `synchronized`.